synapseclient/client.py

```diff
@@ -621,6 +621,76 @@ def invalidateAPIKey(self):
         if self._is_logged_in():
             self.restDELETE("/secretKey", endpoint=self.authEndpoint)
 
+    @functools.lru_cache()
+    def get_user_profile(
+        self,
+        id: Union[str, UserProfile, TeamMember] = None,
+        sessionToken: str = None,
+    ) -> UserProfile:
+        """
+        Get the details about a Synapse user.
+        Retrieves information on the current user if 'id' is omitted or is empty string.
+        :param id:           The userName, UserProfile, or TeamMember of a user
+        :param sessionToken: The session token to use to find the user profile
+        :returns: The user profile for the user of interest.
+
+        Example::
+            my_profile = syn.get_user_name_profile()
+            freds_profile = syn.get_user_name_profile('fredcommo')
+        """
+        if id:
+            if isinstance(id, collections.abc.Mapping) and "ownerId" in id:
+                id = id.ownerId
+            elif isinstance(id, TeamMember):
+                id = id.member.ownerId
+            elif isinstance(id, str):
+                principals = self._findPrincipals(id)
+                if len(principals) == 1:
+                    id = principals[0]["ownerId"]
+                else:
+                    for principal in principals:
+                        if principal.get("userName", None).lower() == id.lower():
+                            id = principal["ownerId"]
+                            break
+                    else:  # no break
+                        raise ValueError('Can\'t find user "%s": ' % id)
+            else:
+                raise TypeError("id must be a string, UserProfile, or TeamMember")
+        else:
+            id = ""
+        uri = f"/userProfile/{id}"
+        return UserProfile(
+            **self.restGET(
+                uri, headers={"sessionToken": sessionToken} if sessionToken else None
+            )
+        )
+
+    @functools.lru_cache()
+    def get_user_profile_by_id(
```

Comment:
I might change this to be `get_user_profile_by_owner_id` to be the most specific . The reason is that `id` just means identifier, and username is an identifier - But we aren't intending for them to call this function with the username.

Reply:
@BryanFauble one thing that we have to think about is that client users generally don't have an understanding of what an "owner_id" is. That said, I do like having being consistent with the API.

I'm slightly confused as why they named it `ownerId`.

Reply:
Yeah that makes sense. The changes here do work but I want to be clear for those that haven't worked in synapse before. Do you have an opinion on renaming or keeping? I'm ok with either.

Reply:
Given what we talked about in sprint review what this is should be fine since there is little understanding what ownerId is